FileSystem/FileMapperMocker.js

```diff
@@ -10,28 +25,490 @@ var USE_FILEMAPPER_MOCKER = true;
+    copyFolderUtil(folder, destination);
+    folder.setTrashed(true);
+  }
+}
+
+/**
+ * Copy file API
+ * API copies all files to the target folder.
+ * @param {Array} sourceFilePaths List of local files to be copied
+ * @param {string} targetFolderPath Destination folder path
+ */
+function copyFile(sourceFilePaths, targetFolderPath) {
+  var targetFolderId;
+  // Get folder, if it doesn't exist then create it
+  try {
+    targetFolderId = this.getFolderId(targetFolderPath);
+  } catch (e) {
+    targetFolderId = this.createFolder(targetFolderPath);
+  }
+  var targetFolder = DriveApp.getFolderById(targetFolderId);
+  // Copy all files
+  for (var i = 0; i < sourceFilePaths.length; i++) {
+    var fileId = this.getFileId(sourceFilePaths[i]);
+    var file = DriveApp.getFileById(fileId);
+    file.makeCopy(file.getName(), targetFolder);
+  }
+}
+
+/**
+ * Copy folder API
+ * API copies all folders to the target folder. Apps Script does not have a
+ * direct copy folder API, folder copy is done recursively
+ * @param {Array} sourceFilePaths List of local folders to be copied
+ * @param {string} targetFolderPath Destination folder path
+ */
+function copyFolder(sourceFilePaths, targetFolderPath) {
+  var targetFolderId;
+  // Get folder, if it doesn't exist then create it
+  try {
+    targetFolderId = this.getFolderId(targetFolderPath);
+  } catch (e) {
+    targetFolderId = this.createFolder(targetFolderPath);
+  }
+  var targetFolder = DriveApp.getFolderById(targetFolderId);
+  // Copy all folders
+  for (var i = 0; i < sourceFilePaths.length; i++) {
+    var folderId = this.getFolderId(sourceFilePaths[i]);
+    var folder = DriveApp.getFolderById(folderId);
+    var destination = targetFolder.createFolder(folder.getName());
+    copyFolderUtil(folder, destination);
+  }
+}
+
+/**
+ * Recursively copy source folder to target folder
+ * @param {Folder} source Source Folder
+ * @param {Folder} target Target Folder
+ */
+function copyFolderUtil(source, target) {
+  var folders = source.getFolders();
+  var files = source.getFiles();
+  while (files.hasNext()) {
+    var file = files.next();
+    file.makeCopy(file.getName(), target);
+  }
+  while (folders.hasNext()) {
+    var subFolder = folders.next();
+    var folderName = subFolder.getName();
+    var targetFolder = target.createFolder(folderName);
+    copyFolderUtil(subFolder, targetFolder);
+  }
+}
+
+/**
+ * Find files by pattern API
+ * This API searches for files which match the given pattern consisting
+ * of wildcards. Wildcard ? matches with any one valid character.
+ * Wildcard * matches with one or more valid characters
+ * @param {string} localPath Local file path of the file
+ * @return {Array} List of files matching the pattern
+ */
+function findFilesByPattern(localPath) {
+  // Get platform agnostic relative path split
+  var pathSplit = this.getRelativePathSplit(localPath);
+  // Get base directory
+  var root = this.getBaseDriveFolder();
+  // Get parent directory of file
```

Comment:
395-400 code is used repeatedly, create a helper function.

Reply:
Fixed: Refactored to using getFinalFolderUtil()